sycl/source/detail/device_impl.cpp

```diff
@@ -314,6 +314,10 @@ bool device_impl::has(aspect Aspect) const {
     return get_info<info::device::usm_restricted_shared_allocations>();
   case aspect::usm_system_allocations:
     return get_info<info::device::usm_system_allocations>();
+  case aspect::ext_intel_pci_device_id:
+    return getPlugin().call_nocheck<detail::PiApiKind::piDeviceGetInfo>(
+               MDevice, PI_DEVICE_INFO_PCI_DEVICE_ID, 0,
+               nullptr, &return_size) == PI_SUCCESS;
```

Comment:
> [cl_intel_attributes_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc#modifications-to-the-opencl-api-specification) says that the corresponding query returns PCI ID only if device has it, i.e. the fact that the extension is supported, doesn't mean that PCI ID exists for the device:
I don't see the issue here. The proposed `ext_intel_pci_device_id` does not claim that it would return such ID for any device. It would return a **PCI ID**. If a device is not a PCI device, it's obvious that it shouldn't have a PCI ID. This is what [cl_intel_attributes_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc#modifications-to-the-opencl-api-specification) says.

> Unfortunately, [deviceId](https://spec.oneapi.io/level-zero/latest/core/api.html#_CPPv4N22ze_device_properties_t8deviceIdE) in Level Zero is poorly documented and it is not clear how returned 32 bits are filled and with which exact data. My concern here is that returned value might not have a uniform format across all backends. See my other comment on the spec extension document with other comments on the topic.
[cl_intel_attributes_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc#modifications-to-the-opencl-api-specification) is clear, and in Level Zero are you aware of any reason why `deviceId` could suddenly change? That structure also contains `vendorId`, which together with `deviceId` form the full PCI ID. Why would Level Zero not return the actual device ID, and what would be the point of not returning it in the lower 16 bits?

Reply:
> I don't think that this should block the PR, but I think that we do need to add a FIXME comment here, because strictly speaking, this implementation (of both has and get_info) is not entirely correct.

What needs to be fixed specifically?

Reply:
> I don't see the issue here. The proposed `ext_intel_pci_device_id` does not claim that it would return such ID for any device. It would return a **PCI ID**. If a device is not a PCI device, it's obvious that it shouldn't have a PCI ID. This is what [cl_intel_attributes_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc#modifications-to-the-opencl-api-specification) says. If it's a PCI device, why wouldn't it have a PCI device ID? 

With that I agree. The inconsistency I'm highlighting here is that I don't see us checking the result returned by `CL_DEVICE_ID_INTEL` before saying whether or not the corresponding aspect is supported. I also don't see us raising an exception in case `CL_DEVICE_ID_INTEL` did not return a PCI ID (i.e. any bit except for low 16 bits is set to non-zero value).

> How is this any different from the existing `ext_intel_pci_address`?

Querying PCI address is not currently supported for OpenCL backend.

https://github.com/intel/llvm/blob/9717cc56cfead3a0045e17a8b1882ed312ede46b/sycl/plugins/opencl/pi_opencl.cpp#L263-L267

Depending on how the corresponding OpenCL extension will be worded (or perhaps there is already such an extension? I'm just not aware of it) it might or might not have the same implementation inconsistency.


> [cl_intel_attributes_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc#modifications-to-the-opencl-api-specification) is clear, and in Level Zero are you aware of any reason why `deviceId` could suddenly change? That structure also contains `vendorId`, which together with `deviceId` form the full PCI ID. Why would Level Zero not return the actual device ID, and what would be the point of not returning it in the lower 16 bits?

The reason I asked the question is because I'm unfamiliar with PCI ID encoding/representation. Is it documented somewhere as a standard? If so, there is absolutely no problem.

Reply:
> With that I agree. The inconsistency I'm highlighting here is that I don't see us checking the result returned by CL_DEVICE_ID_INTEL before saying whether or not the corresponding aspect is supported. I also don't see us raising an exception in case CL_DEVICE_ID_INTEL did not return a PCI ID (i.e. any bit except for low 16 bits is set to non-zero value).

Even if there wouldn't be any bits in the upper 16 bits, it still wouldn't mean that it's an actual valid PCI ID. There still wouldn't be any guarantee by adding this check. But if you think it would be useful, please add it.

> The reason I asked the question is because I'm unfamiliar with PCI ID encoding/representation. Is it documented somewhere as a standard? If so, there is absolutely no problem.

The PCI encoding is widely documented, e.g. https://man7.org/linux/man-pages/man5/pci.ids.5.html:

> Devices on the PCI bus are identified by a combination of a
> vendor ID (assigned by the PCI SIG) and device ID (assigned by
> the vendor). Both IDs are 16-bit integers and the device itself
> provides no translation to a human-readable string.

SYCL can already report the vendor ID part of the PCI ID (`vendor_id`), and a lot of applications are already relying on it. This extension would enable the device ID part.

Reply:
> Querying PCI address is not currently supported for OpenCL backend.

What happens if the PCI address is queried for a non-PCI Level Zero device?

If there are serious doubts about the reliability of this implementation then the value could be renamed to something like `ext_intel_device_id` and specify the PCI device ID only as a possible but not required value. But if the return value can be pretty much anything, it would not make sense to query it at all.

Since this is an Intel extension, and Intel controls both the Level Zero and OpenCL backends, is there any reason why we could not guarantee the consistency of reporting the PCI device ID?

Reply:
> Even if there wouldn't be any bits in the upper 16 bits, it still wouldn't mean that it's an actual valid PCI ID. There still wouldn't be any guarantee by adding this check.

Isn't that even more problematic? Correct me if I am reading it wrong, but the [cl_intel_device_attribute_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc) extension does not make any guarantees that the returned value is a PCI device ID.
I believe this is what @AlexeySachkov is getting at. The only way we check whether or not we can get a PCI device ID is to check if the OpenCL call returns something, but in reality we do not know if it is a PCI device ID. This means for some devices (especially ones that do not have PCI IDs) this may return some other ID, which we act like is a PCI device ID. Is that right?

Reply:
> What happens if the PCI address is queried for a non-PCI Level Zero device? Or a Level Zero device is required to be a PCI device?

That's a very good question, and it plays into the first question from Alexey. The L0 documentation isn't very clear on that. As a result, how can we make any promises that it is actually the PCI device ID we get out?

> If there are serious doubts about the reliability of this implementation then the value could be renamed to something like `ext_intel_device_id` and specify the PCI device ID only as a possible but not required value. But if the return value can be pretty much anything, it would not make sense to query it at all.
>
> Since this is an Intel extension, and Intel controls both the Level Zero and OpenCL backends, is there any reason why we could not guarantee the consistency of reporting the PCI device ID (while we can guarantee the same for the PCI address on Level Zero)?

Honestly I think `ext_intel_device_id` would be better. You are right that it is an Intel extension, but we still have to rely on how the backend specs are defined. We do not want an implementation where we report that the aspect is supported for some device, then return an invalid value, even if it isn't our implementation. And even if we could limit it to our implementation only and we were to rely on them returning PCI device IDs, because of how the specifications are written that could change at any time and we would suddenly have a faulty extension while the backend implementations would still be correct.
As for why you would query it if the return value could be pretty much anything, I guess that question goes for the L0 and OpenCL equivalents as well. I would at least hope that the device ID stays the same for a given device, so if you know an ID (PCI device ID or otherwise) you could make a check for that known ID in your code and do whatever you'd like to do special for it. That should still work, even if the ID could be anything.

Reply:
> Isn't that even more problematic? Correct me if I am reading it wrong, but the [cl_intel_device_attribute_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc) extension does not make any guarantees that the returned value is a PCI device ID.

My understanding is that it **does** guarantee it, **if** it's a PCI device, which is the only kind of the device this extension cares about.

> Honestly I think ext_intel_device_id would be better.

I would rather say that it would be easier to maintain but, depending on the wording of the specs, it could be much less useful, especially if we would allow to report different values per backend (which would make this extension completely useless). Even if we would rename it to `ext_intel_device_id`, I would still make it mandatory to return the PCI ID **if** the device is a PCI device, and something unspecified otherwise (but I'm not sure what would be the use of that). Which is exactly what [cl_intel_device_attribute_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc) specifies.

> And even if we could limit it to our implementation only and we were to rely on them returning PCI device IDs, because of how the specifications are written that could change at any time and we would suddenly have a faulty extension while the backend implementations would still be correct.

But this is far from being the only value that we're already querying from various runtimes, the same risk applies to all of them. What makes querying the PCI ID so much riskier that all the others? Why do we accept the risk for all other values but not for this one?

A completely vague specification would make this extension unusable. If it might break in the future, then it has to be fixed just like any other bug. Yes, we might suddenly have a faulty extension, which would be just like any other bug which has to be fixed. But if we would have a vaguely specified extension, it wouldn't be a bug anymore, but the applications relying on it would break anyway, with no fix in sight. Then all applications would need to come up with their own workarounds, instead of fixing the issue in a centralized way, in the SYCL implementation.

If we can commit to query the PCI address, why would it be so difficult to correctly report the PCI ID as well? If the Level Zero specs are not clear enough, why not simply clarify them then? Only the specs might require a minor update, but no code change would be necessary. I think this would be the proper solution.

At the end of the day, we need a mechanism to **reliably** query the architecture of a GPU. The PCI ID is an obvious way to do it. If this is not possible to do for some reason, we need something else instead. The question is: would something else be any easier to implement than this extension? Regardless of the chosen method, we urgently need a solution to this problem.

Reply:
> That's a very good question, and it plays into the first question from Alexey. The L0 documentation isn't very clear on that. As a result, how can we make any promises that it is actually the PCI device ID we get out?

Only the specs could guarantee the **intent**, the implementation could be faulty regardless. But if the same issue applies to the PCI address, why was that approved to be part of this extension, while adding the PCI ID is currently blocked because of this?

Reply:
> My understanding is that it does guarantee it, if it's a PCI device, which is the only kind of the device this extension cares about.

Then the aspect shouldn't get defined for cases the extension doesn't "care about". The aspect is meant to be a promise that a certain feature is supported on the device. If the user checks if a device has it and it returns `true` then the following query should give a value as defined per this new query extension, which currently says it should be a PCI device ID.

> Even if we would rename it to `ext_intel_device_id`, I would still make it mandatory to return the PCI ID if the device is a PCI device, and something unspecified otherwise (but I'm not sure what would be the use of that). Which is exactly what [cl_intel_device_attribute_query](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_device_attribute_query.asciidoc) specifies.

Absolutely! That seems like a reasonable requirement that the OpenCL extension seems to agree with. The problem here boils down to the fact that this extension is currently promising more than the used OpenCL query does, without doing any reliable checks or filtering for values that do not adhere to that promise (i.e. values that are not PCI device IDs).

> Only the specs could guarantee the intent, the implementation could be faulty regardless. But if the same issue applies to the PCI address, why was that approved to be part of this extension, while adding the PCI ID is currently blocked because of this?

I am okay with assuming that the intent of the L0 specification is to return the PCI device ID. It does mention the PCI configuration after all.

> A completely vague specification would make this extension unusable. If it might break in the future, then it has to be fixed just like any other bug. Yes, we might suddenly have a faulty extension, which would be just like any other bug which has to be fixed. But if we would have a vaguely specified extension, it wouldn't be a bug anymore, but the applications relying on it would break anyway, with no fix in sight. Then all applications would need to come up with their own workarounds, instead of fixing the issue in a centralized way, in the SYCL implementation.

Amending the L0 specification is not the big problem here. The issue is that the OpenCL extension used here is very clear that it may not return a PCI device ID. As such, we cannot just assume that if the query works we will also be able to return a PCI device ID. What I was trying to convey is that any of our OpenCL implementations would be in their full right to return something (now or in the future) that is not a PCI device ID, but with the current version of this SYCL extension and implementation we would accept it as if it was a PCI device ID. Likewise we could have another OpenCL implementation report that it supports the extension query and not return a PCI device ID, and since we would report the device as having the new aspect we would be considering it as supported for this new SYCL extension. Since it isn't a bug in the OpenCL implementations, it is a bug in this PR's implementation/extension.

An option is to ask for a new OpenCL query that guarantees a PCI device ID if it returns a value, but until we have that new query we would have to consider OpenCL backends as not supporting this extension.

> But this is far from being the only value that we're already querying from various runtimes, the same risk applies to all of them. What makes querying the PCI ID so much riskier that all the others? Why do we accept the risk for all other values but not for this one?

Other queries do not use the same L0 properties and the properties they use may be clearer, such as the [`address` member of `zes_pci_properties_t`](https://spec.oneapi.io/level-zero/latest/sysman/api.html#_CPPv420zes_pci_properties_t) for PCI addresses. Again, I am okay with assuming that L0 always returns a PCI device ID as part of `deviceId`, but the bigger problem here is not L0 but OpenCL. The rest of the extension queries in the SYCL extension document this PR touches are only supported on L0 and return `PI_ERROR_INVALID_VALUE` on OpenCL.

Reply:
> Absolutely! That seems like a reasonable requirement that the OpenCL extension seems to agree with. The problem here boils down to the fact that this extension is currently promising more than the used OpenCL query does, without doing any reliable checks or filtering for values that do not adhere to that promise (i.e. values that are not PCI device IDs).

Then the solution is straightforward. Rename the aspect to `ext_intel_device_id` and copy the specs from the OpenCL extension, which is the following:

> A unique device identifier for the OpenCL device.
> 
> If the implementation is driven primarily by a PCI device with a PCI device ID, the low 16 bits of the device identifier must contain that PCI device ID, and the remaining bits must be set to zero. Otherwise, the choice of what to return may be dictated by operating system or platform policies - but should uniquely identify both the device version and any major configuration options (for example, core count in the case of multi-core devices).

Thus the implementation of `ext_intel_device_id` would be perfect for both L0 and OpenCL. Do you have any objections?

Reply:
Sounds good to me! 🚀

Reply:
If @AlexeySachkov still has concerns regarding which exact bits contain the PCI device ID in the value returned by L0 (but I don't see any reason why it would be bit-shifted), then I would be okay with the following relaxed wording:

> A unique device identifier for the device.
> 
> If the implementation is driven primarily by a PCI device with a PCI device ID, the device identifier must be that PCI device ID. Otherwise, the choice of what to return may be dictated by operating system or platform policies - but should uniquely identify both the device version and any major configuration options (for example, core count in the case of multi-core devices).

@AlexeySachkov , do you find this acceptable?

Reply:
Renamed to ext_intel_device_id, updated the specs.